core/src/main/java/org/cru/contentscoring/core/servlets/CopyScoresToTagsServlet.java

```diff
@@ -0,0 +1,259 @@
+                "Unauthorized attempt to move scores to tags by {}",
+                principal == null ? "Anonymous" : principal.getName());
+            response.sendError(401, "You are not authorized to perform this command.");
+            return;
+        }
+
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        Resource root = resourceResolver.getResource(path);
+        if (root == null) {
+            response.sendError(400, "Invalid path");
+            return;
+        }
+
+        TagManager tagManager = resourceResolver.adaptTo(TagManager.class);
+
+        List<Hit> results = findPagesWithScoreProperty(root);
+
+        try {
+            moveScoresToTags(results, tagManager, resourceResolver);
+        } catch (RepositoryException | ReplicationException e) {
+            LOG.error(e.getMessage());
+            response.sendError(500, e.getMessage());
+        }
+    }
+
+    private List<Hit> findPagesWithScoreProperty(final Resource root) {
+        SimpleSearch search = root.adaptTo(SimpleSearch.class);
+        search.addPredicate(buildScorePredicate());
+        search.addPredicate(buildTypePredicate());
+        search.setSearchIn(root.getPath());
+        search.setHitsPerPage(0L); // 0 means unlimited
+
+        try {
+            SearchResult searchResult = search.getResult();
+            LOG.debug("Search took {} seconds", searchResult.getExecutionTime());
+            return searchResult.getHits();
+        } catch (RepositoryException e) {
+            LOG.error("Failed to search for pages with a score property under path {}", root.getPath(), e);
+        }
+        return Lists.newArrayList();
+    }
+
+    private Predicate buildScorePredicate() {
+        Predicate scorePredicate = new Predicate("score", "property");
+        scorePredicate.set("property", JcrConstants.JCR_CONTENT + "/" + "score");
+        scorePredicate.set("operation", "exists");
+        return scorePredicate;
+    }
+
+    private Predicate buildTypePredicate() {
+        Predicate typePredicate = new Predicate("type", "type");
+        typePredicate.set("type", "cq:Page");
+        return typePredicate;
+    }
+
+    private void moveScoresToTags(
+        final List<Hit> results,
+        final TagManager tagManager,
+        final ResourceResolver resourceResolver) throws RepositoryException, ReplicationException {
+
+        List<Resource> pages = Lists.newArrayList();
+        for (Hit result : results) {
+            Resource page = result.getResource();
+            copyScoreTagToPrimaryExperienceFragment(page, resourceResolver, tagManager);
+            moveScoreToTag(page, tagManager);
+            pages.add(page);
+        }
+
+        replicatePages(pages, resourceResolver);
+    }
+
+    private void moveScoreToTag(final Resource page, final TagManager tagManager) throws RepositoryException {
+        Resource pageContent = getJcrContent(page);
+
+        if (pageContent != null) {
```

Comment:
Can `pageContent` even be null? It seems that the query guarantees it exists.

Reply:
It can be null, for example `jcr:content` itself would return null.

Reply:
How can `jcr:content/score` be not null (which seems to be what the query filter requires) and `jcr:content` be null?

Reply:
Ah yes, in this spot you're probably right. I was thinking on line 179 with the experience fragments.

Reply:
Yeah, 179 probably needs the check. 
158 probably doesn't, since it's the same as this (144) check.